### PR TITLE
feat(engine): Implement scheduled workflows

### PR DESCRIPTION
**Pull request by topher-lo**

### CODE REVIEW

File: tracecat/scheduler.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    next_run = croniter(schedule.cron, now).get_next(datetime)
+    return next_run < now + timedelta(seconds=TRACECAT__SCHEDULE_INTERVAL_SECONDS)
+
+
+@retry(
+    wait=wait_exponential(multiplier=1, min=4, max=10),
+    stop=stop_after_attempt(3),
+)
+async def start_workflow(workflow_id):
+    async with httpx.AsyncClient() as client:
```
Comment (daryllimyt) [inline tracecat/scheduler.py]: should probably use AuthenticatedRunnerClient

File: tracecat/scheduler.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        response.raise_for_status()
+        return response
+
+
+async def run_scheduled_workflows():
+    while True:
+        now = datetime.now()
+        with Session(engine) as session:
+            schedules = session.exec(WorkflowSchedule).all()
+            async with httpx.AsyncClient(
```
Comment (daryllimyt) [inline tracecat/scheduler.py]: same as above

File: tracecat/types/schedules.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from pydantic import BaseModel
+
+
+class WorkflowScheduleParam(BaseModel):
```
Comment (daryllimyt) [inline tracecat/types/schedules.py]:
```suggestion
class WorkflowScheduleParams(BaseModel):
```

File: tracecat/scheduler.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    return {"message": "Hello world. I am the Scheduler."}
+
+
+@app.get("/health")
+def check_health() -> dict[str, str]:
+    return {"message": "Hello world. I am the Scheduler. This is the health endpoint."}
+
+
+@app.get("/workflows/{workflow_id}/schedules", response_model=list[WorkflowSchedule])
+def list_schedules(role: Annotated[Role, Depends(authenticate_user)], workflow_id: str):
```
Comment (daryllimyt) [inline tracecat/scheduler.py:189]: does this accept M2M connections? if so use `authenticate_user_or_service` or `authenticate_service`

File: tracecat/scheduler.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+from tracecat.logger import standard_logger
+from tracecat.types.schedules import WorkflowScheduleParam
+
+logger = standard_logger("scheduler")
+
+
+engine: Engine
+
+
+async def should_run_schedule(schedule, now):
```
Comment (daryllimyt) [inline tracecat/scheduler.py]: type hints

**Review by daryllimyt — CHANGES_REQUESTED** [review]
small changes

File: docker-compose.yaml [hunk 6]
```diff
@@ -57,6 +58,30 @@ services:
...
+      TRACECAT__RUNNER_URL: ${TRACECAT__RUNNER_URL}
+      OPENAI_API_KEY: ${OPENAI_API_KEY}
+      SUPABASE_PSQL_URL: ${SUPABASE_PSQL_URL}
+      LOG_LEVEL: ${LOG_LEVEL}
+    container_name: scheduler
+    restart: unless-stopped
+    depends_on:
+      - nginx
+    networks:
+      - scheduler-network
```
Comment (daryllimyt) [inline docker-compose.yaml]:
```suggestion
      - internal
```
or something else

File: tests/test_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+from tracecat.config import TRACECAT__RUNNER_URL
+from tracecat.db import TRACECAT__DB_URI, Workflow, WorkflowSchedule
+from tracecat.scheduler import app, engine, start_scheduler
+
+TEST_SCHEDULER_INTERVAL_SECONDS = 10
+TEST_WORKFLOW_RUN_TIMEOUT = 40  # seconds
+TEST_USER_ID = "3f1606c4-351e-41df-acb4-fb6e243fd071"
+TEST_OTHER_USER_ID = "83de065f-e933-4a69-8a7a-2d796238575d"
+
+os.environ["TRACECAT__DB_ENCRYPTION_KEY"] = Fernet.generate_key().decode()
```
Comment (daryllimyt) [inline tests/test_scheduler.py:37]: nit: optionally put this in the fixture?

File: tracecat/db.py [hunk 8]
```diff
@@ -153,6 +162,22 @@ class WorkflowRun(Resource, table=True):
     action_runs: list["ActionRun"] | None = Relationship(back_populates="workflow_run")
 
 
+class WorkflowSchedule(Resource, table=True):
```
Comment (daryllimyt) [inline tracecat/db.py:165]: test this with fkey constraints enabled. sqlite doesn't have this on by default

**Review by daryllimyt — CHANGES_REQUESTED** [review]
LGTM, minor changes

File: tracecat/db.py (hunk 8 above)
Comment (topher-lo) [reply tracecat/db.py:165]: Later

**Pull request merged**